callouts/java/service-callout/src/main/java/service/ServiceCallout.java

```diff
@@ -0,0 +1,286 @@
+        ProcessingResponse.Builder builder = ProcessingResponse.newBuilder();
+
+        switch (request.getRequestCase()) {
+            case REQUEST_HEADERS:
+                Optional<HeadersOrImmediateResponse> response = onRequestHeaders(request.getRequestHeaders());
+                if (response.isPresent()) {
+                    HeadersOrImmediateResponse headersOrImmediateResponse = response.get();
+                    if (headersOrImmediateResponse.isImmediateResponse()) {
+                        builder.setImmediateResponse(headersOrImmediateResponse.getImmediateResponse());
+                    } else {
+                        builder.setRequestHeaders(headersOrImmediateResponse.getHeadersResponse());
+                    }
+                }
+                break;
+            case RESPONSE_HEADERS:
+                onResponseHeaders(builder.getResponseHeadersBuilder(), request.getResponseHeaders());
+                break;
+            case REQUEST_BODY:
+                onRequestBody(builder.getRequestBodyBuilder(), request.getRequestBody());
+                break;
+            case RESPONSE_BODY:
+                onResponseBody(builder.getResponseBodyBuilder(), request.getResponseBody());
+                break;
+            case REQUEST_TRAILERS:
+                break;
+            case RESPONSE_TRAILERS:
+                break;
+            case REQUEST_NOT_SET:
+            default:
+                logger.log(Level.WARNING, "Received a ProcessingRequest with no request data.");
+                break;
+        }
+
+        return builder.build();
+    }
+
+    /**
+     * Callback method invoked upon receiving request headers.
+     *
+     * @param headers Incoming request headers
+     * @return Optional HeadersOrImmediateResponse
+     */
+    public Optional<HeadersOrImmediateResponse> onRequestHeaders(HttpHeaders headers) {
+        return Optional.empty();
+    }
+
+    /**
+     * Callback method invoked upon receiving response headers.
+     *
+     * @param headerResponse Builder for modifying response headers
+     * @param headers        Incoming response headers
+     */
+    public void onResponseHeaders(HeadersResponse.Builder headerResponse, HttpHeaders headers) {
+        // Default implementation does nothing
+    }
+
+    /**
+     * Callback method invoked upon receiving request body.
+     *
+     * @param bodyResponse Builder for modifying response body
+     * @param body         Incoming request body
+     */
+    public void onRequestBody(BodyResponse.Builder bodyResponse, HttpBody body) {
+        // Default implementation does nothing
+    }
+
+    /**
+     * Callback method invoked upon receiving response body.
+     *
+     * @param bodyResponse Builder for modifying response body
+     * @param body         Incoming response body
+     */
+    public void onResponseBody(BodyResponse.Builder bodyResponse, HttpBody body) {
+        // Default implementation does nothing
+    }
```

Comment:
This may be a useful distinction in the future but as the server is pretty light on any unique functionality, and as the other languages in this repo use a unified server processor, it is probably a good idea to keep the combined structure for now.